cuda/dot_benchmark.cu

```diff
@@ -0,0 +1,268 @@
+    // Increase in number of elements between consecutive benchmark runs
+    constexpr std::size_t row_incr = 2'000'000;
+    // Number of benchmark runs (ignoring randomization)
+    constexpr std::size_t steps = (max_size - start) / row_incr;
+    // Number of benchmark restarts with a different randomization for vectors
+    // Only used for a detailed error run
+    constexpr std::size_t max_randomize_num{10};
+
+    std::vector<std::size_t> benchmark_vec_size((steps + 1));
+    std::vector<double> benchmark_time((steps + 1) * benchmark_num);
+    // std::vector<value_type> benchmark_error((steps + 1) * benchmark_num);
+    // stores the result for all different benchmark runs to compute the error
+    const auto actual_randomize_num = detailed_error ? max_randomize_num : 1;
+    std::vector<value_type> raw_result(actual_randomize_num * (steps + 1) *
+                                       benchmark_num);
+    const auto get_raw_idx = [benchmark_num, actual_randomize_num](
+                                 std::size_t rnd, std::size_t step,
+                                 std::size_t bi) {
+        return step * actual_randomize_num * benchmark_num +
+               bi * actual_randomize_num + rnd;
+    };
+
+    for (std::size_t randomize = 0; randomize < actual_randomize_num;
+         ++randomize) {
+        if (randomize != 0) {
+            write_random({{max_size, 1}}, vector_dist, rengine,
+                         ar_data.cpu_x_nc());
+            write_random({{max_size, 1}}, vector_dist, rengine,
+                         ar_data.cpu_y_nc());
+            ar_data.copy_cpu_to_gpu();
+            st_data.convert_from(ar_data);
+        }
+        for (std::size_t vec_size = start, i = 0; vec_size <= max_size;
+             vec_size += row_incr, ++i) {
+            benchmark_vec_size.at(i) = vec_size;
+            const matrix_info x_info{{vec_size, 1}};
+            const matrix_info y_info{{vec_size, 1}};
+
+            for (std::size_t bi = 0; bi < benchmark_num; ++bi) {
+                const std::size_t idx = i * benchmark_num + bi;
+                auto curr_lambda = [&]() {
+                    std::get<1>(benchmark_info[bi])(x_info, y_info);
+                };
+                benchmark_time.at(idx) =
+                    benchmark_function(curr_lambda, detailed_error);
+                raw_result[get_raw_idx(randomize, i, bi)] =
+                    std::get<2>(benchmark_info[bi])();
+            }
+            // const auto result_ref =
+            //    raw_result[get_raw_idx(randomize, i, benchmark_reference)];
+            // for (std::size_t bi = 0; bi < benchmark_num; ++bi) {
+            //    const std::size_t idx = i * benchmark_num + bi;
+            //    benchmark_error.at(idx) +=
+            //        get_error(raw_result[get_raw_idx(bi)], result_ref);
+            //}
+        }
+    }
+    for (std::size_t i = 0; i <= steps; ++i) {
```

Comment:
It was not so much in terms of reuse but rather in highlighting the main parts as people might look at this file first when starting to dive into the code (core of the benchmark). The main is 200 lines long I was wondering if there are strategies to at least half that. Since the kernels and whatnot are now well documented it might be fine like this.